### PR TITLE
Issue 2

### PR DESCRIPTION
**Pull request by tomharvey**
PR to demonstrate the [issue](https://github.com/tbuehlmann/dry-struct-setters/issues/2) with optional keys and a potential (but maybe not the best) solution to the issue.

### CODE REVIEW

File: lib/dry/struct/setters.rb [hunk 1]
```diff
@@ -27,6 +27,7 @@ def attributes(new_schema)
       end
 
       def self.define_setter_for(struct:, attribute:, type:)
+        attribute = Dry::Struct::Setters.remove_trailling_question_mark(attribute)
```
Comment (tbuehlmann) [inline lib/dry/struct/setters.rb]: No need for `Dry::Struct::Setters` here, `remove_trailling_question_mark(attribute)` will just work.

Comment (tbuehlmann) [reply lib/dry/struct/setters.rb]: And it should be `trailing` with a single `l`, right?

File: lib/dry/struct/setters.rb [hunk 2]
```diff
@@ -39,6 +40,12 @@ def self.define_setter_for(struct:, attribute:, type:)
           end
         end
       end
+
+      def self.remove_trailling_question_mark(attribute)
+        return if attribute.class != Symbol
+        attribute = attribute.to_s.chomp "?"
+        return attribute.to_sym
+      end
```
Comment (tbuehlmann) [inline lib/dry/struct/setters.rb]: Any reason to care about `attribute` being a `Symbol` What about this?

```ruby
def self.remove_trailing_question_mark(attribute)
  attribute.to_s.chomp('?')
end
```

Having that, I probably wouldn't extract this into a method and just do this:

```ruby
def self.define_setter_for(struct:, attribute:, type:)
  attribute = attribute.to_s.chomp('?')
  setter = "#{attribute}=".to_sym

  struct.class_eval do
    unless instance_methods.include?(setter)
      define_method(setter) do |value|
        @attributes[attribute] = type.call(value)
      end

      setter
    end
  end
end
```

What do you think?

File: spec/dry/struct/setters_spec.rb [hunk 3]
```diff
@@ -19,5 +20,9 @@
     it 'dynamically defines setters for attributes defined after the inclusion' do
       expect { subject.after = 'new-after' }.to change(subject, :after).from('after').to('new-after')
     end
+
+    it 'defines setters for imittable attributes' do
```
Comment (tbuehlmann) [inline spec/dry/struct/setters_spec.rb]: Should be `omittable`?

**Review by tbuehlmann — CHANGES_REQUESTED** [review]
Hey, I reviewed the PR and found some minor things.

File: lib/dry/struct/setters.rb (hunk 2 above)
Comment (tomharvey) [reply lib/dry/struct/setters.rb]: Returning if the attribute isn't a symbol was to handle a separate thing ... I might raise as an issue to better explain that "thing", but will remove from here.

File: spec/dry/struct/setters_spec.rb (hunk 3 above)
Comment (tomharvey) [reply spec/dry/struct/setters_spec.rb]: Yeesh! Who's fingers are these?! There are more typo-s in this PR than LOC!

File: lib/dry/struct/setters.rb (hunk 2 above)
Comment (tomharvey) [reply lib/dry/struct/setters.rb]: I extracted it into its own method for readability. Seemed like the reasoning behind `chomp(?)` would be difficult to work out without some awareness of the issue. I've added a link to the issue for future readers of the code.

**Comment by tomharvey** [comment]
Thanks for the feedback. I've made the changes suggested except:

Kept it as a separate method to provide a place to explain why we would want to chomp a ?

And, I still symbolize after the chomping. If I don't, the test fails. No idea why, given the very next line is string concatenation and symbolisation!? Sorry - I would dig into what's going on there, but my first child was born last week, so my priorities have changed since I first created this PR and my level of sleep is ... 💤 

I guess I could justify it that the method should *only* remove the ?, not change the type. So, symbol in, symbol out.

File: lib/dry/struct/setters.rb [hunk 4]
```diff
@@ -39,6 +40,11 @@ def self.define_setter_for(struct:, attribute:, type:)
           end
         end
       end
+
+      def self.remove_trailing_question_mark(attribute)
+        #  See https://github.com/tbuehlmann/dry-struct-setters/issues/2
+        return attribute.to_s.chomp('?').to_sym
```
Comment (tbuehlmann) [inline lib/dry/struct/setters.rb]: No need for the `return` here. Once this is done, we're good to merge.

**Comment by tbuehlmann** [comment]
> but my first child was born last week

First of all: Congratulations!

About the `to_sym`: My bad, didn't see `attribute` is used a few lines below:

```@attributes[attribute] = type.call(value)```

And for that, it's important for `attribute` to be a `Symbol`.

**Review by tbuehlmann — APPROVED** [review]

**Comment by tbuehlmann** [comment]
I just released [0.3.0](https://rubygems.org/gems/dry-struct-setters/versions/0.3.0) with dry-struct >= 1.0.0 and your additions. Thanks!

**Pull request merged**